resources/type-system/flow-analysis-v0.md

```diff
@@ -0,0 +1,339 @@
+  nodes after a control flow split which are only reachable by one path through
+  the split, and is defined as `FlowModel(r1, VM)` where `r0 = pop(r)` and `r1 =
+  push(pop(r0), top(r0) && top(r1))`.  Equivalently, `drop(M)` may be thought of
+  as `join(M, M)`.
+
+- `join(M1, M2)`, where `M1` and `M2` are flow models, represents the union of
+  two flow models and is defined as follows:
+
+  We define `join(M1, M2)` where `M1 = FlowModel(r1, VI1)` and `M2 =
+  FlowModel(r2, VI2))` and `pop(r1) = pop(r2) = r0` for some `r0` to be `M3` where:
+    - if `top(r1)` is true and `top(r2)` is false, then `M3` is `FlowModel(r0, VI1)`.
+    - if `r1` is false and `r2` is true, then `M3` is `FlowModel(pop(r2), VI2)`.
+    - otherwise `M3 = FlowModel(r3, VI3)` where:
+      - `r3` is `push(pop(r0), top(r0) && top(r1) && top(r2))`
+      - `VI3` maps each variable `v` in the domain of `VI1` and `VI2` to
+      `joinV(VI1(v), VI2(v))`.  Note that any variable which is in domain of
+      only one of the two is dropped, since it is no longer in scope.
+
+  Both join and joinV are commutative and associative by construction.
+
+  For brevity, we will sometimes extend `join` to more than two arguments in the
+  obvious way.  For example, `join(M1, M2, M3)` represents `join(join(M1, M2),
+  M3)`, and `join(S)`, where S is a set of models, denotes the result of folding
+  all models in S together using `join`.
+
+- `exit(M)` represents the model corresponding to a program location which is
+  unreachable, but is otherwise modeled by flow model `M = FlowModel(r, VI)`,
+  and is defined as `FlowModel(push(pop(r), false), VI)`
+
+### Promotion
+
+Promotion policy is defined by the following operations on flow models.
+
+Policy:
+  - We say that at type `T` is a type of interest for a variable `x` in a list
+    of test sites `sites` if `sites` contains `x is T` or `x as T`.
+
+  - We say that a variable `x` is promotable via initialization given variable
+    model `VM` if:
+    - `VM = VariableModel(declared, promoted, sites, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and `promoted` is empty
+    - and `x` is declared with no explicit type and no initializer
+    - and `assigned` is false and `unassigned` is true
+
+  - We say that a variable `x` is promotable via assignment of an expression of
+    type `T` given variable model `VM` if
+    - `VM = VariableModel(declared, promoted, sites, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and `promoted = S::l` or (`promoted = []` and `declared = S`)
+    - and `T <: S` and not `S <: T`
+    - and `T` is a type of interest for `x` in `sites`
+
+  - We say that a variable `x` is demotable via assignment of an expression of
+    type `T` given variable model `VM` if
+    - `VM = VariableModel(declared, promoted, sites, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and promoted contains `T`
```

Comment:
Allow demotion to anything that we had previously promoted to?  Probably also need to include the declared type here.